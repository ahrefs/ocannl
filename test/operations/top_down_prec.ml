(* A simple demo of using the Threefry4x32 PRNG in OCANNL *)

open Base
module Tensor = Ocannl.Tensor
module Train = Ocannl.Train
module TDSL = Ocannl.Operation.TDSL
module Tn = Ir.Tnode

let () =
  Tensor.unsafe_reinitialize ();
  let module Backend = (val Backends.fresh_backend ()) in
  let%op d = ("a" [ 2 ] + "b" [ 2 ]) *. "c" [ 2 ] in
  Tn.update_prec b.value Ir.Ops.half;
  Tn.update_prec d.value Ir.Ops.bfloat16;
  (* Even when the default precision is single, c is bfloat16 and a is half. *)
  Ocannl.Train.set_hosted d.value;
  ignore (Ocannl.Train.forward_once (module Backend) d);
  Train.printf_tree d
