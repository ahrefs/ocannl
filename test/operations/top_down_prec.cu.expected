#include <cuda_fp16.h>
#include <cuda_bf16.h>
extern "C" __global__ void d_fwd(
    __half *a,
    __half *b,
    __nv_bfloat16 *c,
    __nv_bfloat16 *d) {
  /* FIXME: single-threaded for now. */if (threadIdx.x != 0 || blockIdx.x != 0) { return; };

  /* Local declarations and initialization. */

  /* Main logic. */
  /* d fwd */
  d[0] = (((__nv_bfloat16)(a[0]) + (__nv_bfloat16)(b[0])) * c[0]);
  /* end */
}